### PR TITLE
[stdlib] Guard for empty singleton in dictionary builder

### PR DESCRIPTION
**Pull request by Azoy**
In preparation of making the empty singletons read only, we need to guard against attempts to write to values in this object. This dictionary builder used by Foundation unconditionally writes to the count based on the initialized count returned from the closure. If the capacity is 0, then we should just skip building the dictionary.

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please smoke test

**Review by Catfish-Man — APPROVED** [review]
lgtm, excited about this change :)

File: stdlib/public/core/DictionaryBuilder.swift [hunk 1]
```diff
@@ -102,6 +102,13 @@ extension _NativeDictionary {
     ) -> Int
   ) {
     self.init(capacity: capacity)
+
+    // If the capacity is 0, then our storage is the empty singleton. Those are
+    // read only, so we shouldn't attempt to write to them.
+    if capacity == 0 {
+      return
```
Comment (lorentey) [inline stdlib/public/core/DictionaryBuilder.swift:109]: Marginal nit: We should still call the function argument exactly once, even if `capacity` happens to be zero.

```suggestion
      let c = initializer(
        UnsafeMutableBufferPointer(start: nil, count: 0), 
        UnsafeMutableBufferPointer(start: nil, count: 0))
      _precondition(c == 0)
      return
```

**Review by lorentey — APPROVED** [review]

**Comment by Azoy** [comment]
@swift-ci please smoke test

**Comment by Azoy** [comment]
@swift-ci please smoke test

**Pull request merged**